### PR TITLE
tests: switch to python_with_gil

### PR DESCRIPTION
**Pull request by mejrs**
This replaces most uses of `acquire_gil` with `with_gil`.

This is only in /src, not in /tests (yet)

### CODE REVIEW

File: src/python.rs [hunk 1]
```diff
@@ -805,4 +803,24 @@ mod tests {
         assert!(PythonVersionInfo::from_str("3.5.2a1+") < (3, 6));
         assert!(PythonVersionInfo::from_str("3.5.2a1+") > (3, 4));
     }
+
+    #[test]
+    fn test_acquire_gil() {
```
Comment (davidhewitt) [inline src/python.rs:808]: Looks like `PyGILState_Check` requires the unlimited api. (See #1787)

```suggestion
    #[test]
    #[cfg(not(Py_LIMITED_API))]
    fn test_acquire_gil() {
```

File: src/types/iterator.rs [hunk 2]
```diff
@@ -107,65 +107,62 @@ mod tests {
...
     fn iter_refcnt() {
-        let obj;
-        let count;
-        {
-            let gil_guard = Python::acquire_gil();
-            let py = gil_guard.python();
-            obj = vec![10, 20].to_object(py);
-            count = obj.get_refcnt(py);
-        }
+        let (obj, count) = Python::with_gil(|py| -> (Py<PyAny>, isize) {
```
Comment (davidhewitt) [inline src/types/iterator.rs:122]: Looks like you'll either need to tweak imports for abi3 or remove the closure return type.

```suggestion
        let (obj, count) = Python::with_gil(|py| {
```

**Review by davidhewitt — APPROVED** [review]
This is heroic! Thank you so much for doing this, I'm inspired to open a PR to deprecate `acquire_gil` now 😄 

There's a few CI issues, which I think I've managed to point out below...

File: src/types/iterator.rs (hunk 2 above)
Comment (mejrs) [reply src/types/iterator.rs:122]: Right that probably should have been a c_int anyway

**Review by davidhewitt — APPROVED** [review]
I think that the PyPy failures were independently fixed by #1787 (which I just merged), so I'm going to merge this now and worry about CI if stuff still breaks 😅 

Thanks again for this tidy-up!

**Pull request merged**